src/caffe/test/test_inner_product_layer.cpp

```diff
@@ -91,6 +135,79 @@ TYPED_TEST(InnerProductLayerTest, TestForward) {
   }
 }
 
+/**
+ * @brief Init. an IP layer without transpose + random weights,
+ * run Forward, save the result.
+ * Init. another IP layer with transpose.
+ * manually copy and transpose the weights from the first IP layer,
+ * then run Forward on the same input and check that the result is the same
+ */
+TYPED_TEST(InnerProductLayerTest, TestForwardTranspose) {
+  typedef typename TypeParam::Dtype Dtype;
+  this->blob_bottom_vec_.push_back(this->blob_bottom_);
+  bool IS_VALID_CUDA = false;
+#ifndef CPU_ONLY
+  IS_VALID_CUDA = CAFFE_TEST_CUDA_PROP.major >= 2;
+#endif
+  if (Caffe::mode() == Caffe::CPU ||
+      sizeof(Dtype) == 4 || IS_VALID_CUDA) {
+    LayerParameter layer_param;
+    InnerProductParameter* inner_product_param =
+        layer_param.mutable_inner_product_param();
+    inner_product_param->set_num_output(10);
+    inner_product_param->mutable_weight_filler()->set_type("uniform");
+    inner_product_param->mutable_bias_filler()->set_type("uniform");
+    inner_product_param->mutable_bias_filler()->set_min(1);
+    inner_product_param->mutable_bias_filler()->set_max(2);
+    inner_product_param->set_transpose(false);
+    shared_ptr<InnerProductLayer<Dtype> > layer(
+        new InnerProductLayer<Dtype>(layer_param));
+    layer->SetUp(this->blob_bottom_vec_, this->blob_top_vec_);
+    layer->Forward(this->blob_bottom_vec_, this->blob_top_vec_);
+    const int count = this->blob_top_->count();
+    Blob<Dtype>* const top = new Blob<Dtype>();
+    top->ReshapeLike(*this->blob_top_);
+    caffe_copy(count, this->blob_top_->cpu_data(), top->mutable_cpu_data());
+    this->blob_top_vec_.clear();
+    this->blob_top_vec_.push_back(new Blob<Dtype>());
+    inner_product_param->set_transpose(true);
+    shared_ptr<InnerProductLayer<Dtype> > ip_t(
+        new InnerProductLayer<Dtype>(layer_param));
+    ip_t->SetUp(this->blob_bottom_vec_, this->blob_top_vec_);
+    const int count_w = layer->blobs()[0]->count();
+    EXPECT_EQ(count_w, ip_t->blobs()[0]->count());
+    // manually copy and transpose the weights from 1st IP layer into 2nd
+    const Dtype* w = layer->blobs()[0]->cpu_data();
+    Dtype* w_t = ip_t->blobs()[0]->mutable_cpu_data();
+    const int WIDTH = layer->blobs()[0]->shape(1);
```

Comment:
Got it, will fix. Thanks for clarifying.
On Feb 3, 2016 23:27, "Jeff Donahue" <notifications@github.com> wrote:

> In src/caffe/test/test_inner_product_layer.cpp
> <https://github.com/BVLC/caffe/pull/3612#discussion_r51797892>:
>
> > +    const int count = this->blob_top_->count();
> > +    Blob<Dtype>* const top = new Blob<Dtype>();
> > +    top->ReshapeLike(*this->blob_top_);
> > +    caffe_copy(count, this->blob_top_->cpu_data(), top->mutable_cpu_data());
> > +    this->blob_top_vec_.clear();
> > +    this->blob_top_vec_.push_back(new Blob<Dtype>());
> > +    inner_product_param->set_transpose(true);
> > +    shared_ptr<InnerProductLayer<Dtype> > ip_t(
> > +        new InnerProductLayer<Dtype>(layer_param));
> > +    ip_t->SetUp(this->blob_bottom_vec_, this->blob_top_vec_);
> > +    const int count_w = layer->blobs()[0]->count();
> > +    EXPECT_EQ(count_w, ip_t->blobs()[0]->count());
> > +    // manually copy and transpose the weights from 1st IP layer into 2nd
> > +    const Dtype* w = layer->blobs()[0]->cpu_data();
> > +    Dtype* w_t = ip_t->blobs()[0]->mutable_cpu_data();
> > +    const int WIDTH = layer->blobs()[0]->shape(1);
>
> In general we follow the Google style guide, but I think our convention is
> to use regular lowercase with underscore variable names (my_variable_name)
> for consts initialized from other variables (as in this case), but camel
> case with k prefix (kMyVariableName) for consts directly initialized with
> their values known at compile time per the Google style guide
> <https://google.github.io/styleguide/cppguide.html#Constant_Names>. We
> use all caps names only for macros AFAIK.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3612/files#r51797892>.
>